tools/compile_tf_graph.py

```diff
@@ -159,34 +210,86 @@ def post_compile(cls, rec_layer_name, network, output_file_name=None):
         "var_op": var.var.op.name,
         "shape": [int(d) if d is not None else None for d in var.var_data_shape.batch_shape],
         "dtype": var.var.dtype.base_dtype.name}
-      if not name.startswith("stochastic_var_"):
-        init_ops.append(var.init_op())
-        tile_batch_ops.append(var.tile_batch_op(tile_batch_repetitions))
-        select_src_beams_ops.append(var.select_src_beams_op(src_beams=src_beams))
-      if not name.startswith("stochastic_var_") and not name.startswith("base_"):
-        next_step_ops.append(var.final_op())
-    info["init_op"] = tf.group(*init_ops, name="rec_step_by_step_init_op").name
-    info["next_step_op"] = tf.group(*next_step_ops, name="rec_step_by_step_update_op").name
-    info["tile_batch"] = {
-      "op": tf.group(*tile_batch_ops, name="rec_step_by_step_tile_batch_op").name,
-      "repetitions_placeholder": tile_batch_repetitions.op.name}
-    info["select_src_beams"] = {
-      "op": tf.group(*select_src_beams_ops, name="rec_step_by_step_select_src_beams_op").name,
-      "src_beams_placeholder": src_beams.op.name}
+      assert var.var not in network.get_saveable_params_list(), 'StateVars are not restorable from the original model check point'
+      if name.startswith("base_"):
+        init_base_ops_list.append(var.init_op())
+        info["collections"]["encode_ops"].append(var.init_op().name)
+
+    # tile batch and group base ops #
+    init_base_ops_list.append( cell.parent_tile_multiples_var.initializer )
+    info["collections"]["encode_ops"].append(cell.parent_tile_multiples_var.initializer.name)
+    init_base_op = tf.group(*init_base_ops_list, name="rec_step_by_step_init_base_op")
+
+    # loop vars #
+    for name, var in sorted(rec_layer.state_vars.items()):
+      if name.startswith("stochastic_var_") or name.startswith("base_"): continue
+      # global vars (hypothesis independent)
+      if name.startswith("global_"):
+        global_vars_coll.append(var.var)
+        info["collections"]["global_vars"].append(var.name)
+      # dependency on init_base_op for each init_op
+      with tf.control_dependencies([init_base_op]):  
+        init_ops_list.append(var.init_op())
+      info["collections"]["encode_ops"].append(var.init_op().name)
+      # choice-dependent loop vars are not I/O stored, but need to be updated after feeding choices
+      if var.choice_dependent:
+        update_ops_list.append(var.final_op())
+        info["collections"]["update_ops"].append(var.final_op().name)
+      # update loop vars after the last decode op
+      elif var.choice_dependent is False:
```

Comment:
Such a bool check is wrong Python style. (Please really check the PyCharm comments, it will tell you that.)